### PR TITLE
Improve durability of fixed point conversion function

### PR DESCRIPTION
**Pull request by maltekliemann**
### What does it do?

Fixes https://github.com/zeitgeistpm/zeitgeist/issues/1131 and generally restructures the function to be more rigid and clear.

### CODE REVIEW

File: primitives/src/math/fixed.rs [hunk 1]
```diff
@@ -243,36 +243,35 @@ impl<F: Fixed + ToString, N: TryFrom<u128>> FromFixedToDecimal<F> for N {
...
+            return Err("The string contains multiple decimal points");
         }
 
-        let (int_part, frac_part) = (parts[0], parts[1]);
-        let mut increment = false;
-
+        let mut increment_int_part = false;
         let new_frac_part = if frac_part.len() < decimals_usize {
             format!("{}{}", frac_part, "0".repeat(decimals_usize.saturating_sub(frac_part.len())))
         } else {
```
Comment (Chralt98) [inline primitives/src/math/fixed.rs:257]:
```suggestion
        } else if frac_part.len() > decimal_usize {
```

> Note that `round_digit` might be `None`, for example if `frac_part = 245` and `decimals = 3`. There's no need to round in that case.

So I assume it's always the case that if `frac_part.len() == decimals` then we don't need to round.

File: primitives/src/math/fixed.rs [hunk 2]
```diff
@@ -243,36 +243,35 @@ impl<F: Fixed + ToString, N: TryFrom<u128>> FromFixedToDecimal<F> for N {
...
-                Some(d) if d >= '5' => increment = true,
-                _ => {}
+            let round_digit = frac_part.chars().nth(decimals_usize).and_then(|c| c.to_digit(10));
+            // Note that `round_digit` might be `None`, for example if `frac_part = 245` and
+            // `decimals = 3`. There's no need to round in that case.
+            if let Some(d) = round_digit {
+                if d >= 5 {
+                    increment_int_part = true;
+                }
             }
```
Comment (Chralt98) [inline primitives/src/math/fixed.rs:265]:
```suggestion
            let round_digit = frac_part.chars().nth(decimals_usize).and_then(|c| c.to_digit(10)).ok_or("The char at decimals_usize was not found, although the frac_part length is higher than decimals_usize.")?;
            if round_digit >= 5 {
                increment_int_part = true;
            }
```

Comment (maltekliemann) [reply primitives/src/math/fixed.rs:265]: Yeah, I had exactly this. But it turns out it's exactly _not_ what want ;) As the comment explains, `round_digit` might be `None`, in which case we can pretend that it's zero; or, equivalently, if it's `None`, then we just don't do any rounding. You'll find that a couple of tests will fail if we make the proposed change.

Comment (maltekliemann) [reply primitives/src/math/fixed.rs:265]: Nevermind, just saw the comment above.

Comment (maltekliemann) [reply primitives/src/math/fixed.rs:265]: Yeah, nice suggestion. The only problem here is that we also need an `else` case, so I think it's fine to leave it the way it is.

**Review by Chralt98 — APPROVED** [review]

**Review by sea212 — APPROVED** [review]

**Pull request merged**